src/Microsoft.Data/PrimitiveColumnContainer.cs

```diff
@@ -69,91 +84,224 @@ public PrimitiveColumnContainer(long length = 0)
                 if (lastBuffer.Length == lastBuffer.MaxCapacity)
                 {
                     lastBuffer = new DataFrameBuffer<T>();
                     Buffers.Add(lastBuffer);
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
                 }
                 int allocatable = (int)Math.Min(length, lastBuffer.MaxCapacity);
                 lastBuffer.EnsureCapacity(allocatable);
+                DataFrameBuffer<byte> lastNullBitMapBuffer = NullBitMapBuffers[NullBitMapBuffers.Count - 1];
+                lastNullBitMapBuffer.EnsureCapacity((int)Math.Ceiling(allocatable / 8.0));
                 lastBuffer.Length = allocatable;
+                lastNullBitMapBuffer.Length = allocatable;
                 length -= allocatable;
                 Length += lastBuffer.Length;
             }
         }
 
-        public void Append(T value)
+        public void Append(T? value)
         {
             if (Buffers.Count == 0)
             {
                 Buffers.Add(new DataFrameBuffer<T>());
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
             DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
             if (lastBuffer.Length == lastBuffer.MaxCapacity)
             {
                 lastBuffer = new DataFrameBuffer<T>();
+                Buffers.Add(lastBuffer);
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
-            lastBuffer.Append(value);
+            lastBuffer.Append(value ?? default);
+            SetValidityBit(Length, value.HasValue ? true : false);
             Length++;
         }
 
+        public void AppendMany(T? value, long count)
+        {
+            if (!value.HasValue)
+            {
+                NullCount += count;
+            }
+            while (count > 0)
+            {
+                if (Buffers.Count == 0)
+                {
+                    Buffers.Add(new DataFrameBuffer<T>());
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
+                }
+                DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
+                if (lastBuffer.Length == lastBuffer.MaxCapacity)
+                {
+                    lastBuffer = new DataFrameBuffer<T>();
+                    Buffers.Add(lastBuffer);
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
+                }
+                int allocatable = (int)Math.Min(count, lastBuffer.MaxCapacity);
+                lastBuffer.EnsureCapacity(allocatable);
+                lastBuffer.Span.Slice(lastBuffer.Length, allocatable).Fill(value ?? default);
+                lastBuffer.Length += allocatable;
+                Length += allocatable;
+
+                DataFrameBuffer<byte> lastNullBitMapBuffer = NullBitMapBuffers[NullBitMapBuffers.Count - 1];
+                int nullBitMapAllocatable = (int)Math.Ceiling(allocatable / 8.0);
```

Comment:
```suggestion
                int nullBitMapAllocatable = (allocatable + 7) / 8;
```
No need for ceil and floatint point.

Reply:
I'm no expert on compilers/runtimes, but I wonder why the compiler doesn't already do this

Reply:
Good question, which I can't answer.
But neither GCC nor Clang do this optimization.